src/types/map.rs

```diff
@@ -12,472 +12,332 @@ use error::Result;
 #[derive(Debug)]
 pub struct Map {
     metadata: ColTypeOption,
-    data: HashMap<String, CBytes>,
+    data: Vec<(CBytes, CBytes)>,
 }
 
 impl Map {
-    /// Creates new `Map` basing on provided data and key and value types.
-    // TODO: Need to return Result<Map>. If key has not string-like type
-    // return Err otherwise return Ok.
+    /// Creates new `Map` using the provided data and key and value types.
     pub fn new(data: Vec<(CBytes, CBytes)>, meta: ColTypeOption) -> Map {
-        let accumulator: HashMap<String, CBytes> = HashMap::new();
-
-        // check that key could be converted into String
-        let serializer = if let Some(ColTypeOptionValue::CMap((ref key_type, _))) = meta.value {
-            match key_type.id {
-                ColType::Custom => decode_custom,
-                ColType::Ascii => decode_ascii,
-                ColType::Varchar => decode_varchar,
-                // unreachable ??
-                // do we need this arm? is it reachable due to the protocol?
-                _ => unimplemented!(),
-            }
-        } else {
-            // do we need this arm? is it reachable due to the protocol?
-            unreachable!();
-        };
-
-        let map: HashMap<String, CBytes> = data.iter()
-            .fold(accumulator, |mut acc, kv| {
-                let (key_b, value_b) = kv.clone();
-                let key: String = serializer(key_b.as_slice()).unwrap();
-
-                acc.insert(key, value_b);
-
-                return acc;
-            });
-
-        return Map {
+        Map {
             metadata: meta,
-            data: map,
-        };
+            data: data,
+        }
     }
 }
 
-// into hash map which values are blobs
-impl AsRust<HashMap<String, Vec<u8>>> for Map {
-    /// Converts `Map` into `HashMap<String, Vec<u8>>` for blob values.
-    fn as_rust(&self) -> Result<HashMap<String, Vec<u8>>> {
-        let map: HashMap<String, Vec<u8>> = HashMap::new();
-
-        match self.metadata.value {
-            Some(ColTypeOptionValue::CMap((_, ref value_type_option))) => {
-                match value_type_option.id {
-                    ColType::Blob => {
-                        Ok(self.data
-                            .iter()
-                            .fold(map, |mut acc, (k, vb)| {
-                                acc.insert(k.clone(), decode_blob(vb.as_plain()).unwrap());
-                                return acc;
-                            }))
-                    }
-                    _ => unreachable!(),
-                }
+macro_rules! data_as_rust {
```

Comment:
It looks as a macros which could be used across the whole `types` module in order to reduce boilerplate in a same fashion as it had been done here. So could you please move this macros into some separate module which we could share inside `types`.